cmd/mdatagen/internal/loader.go

```diff
@@ -373,3 +401,50 @@ func packageName() (string, error) {
 	}
 	return strings.TrimSpace(string(output)), nil
 }
+
+type featureGate struct {
+	// Required.
+	Id string `mapstructure:"id"`
+	// Description describes the purpose of the attribute.
+	Description string `mapstructure:"description"`
+	// Stage current stage at which the feature gate is in the development lifecyle
+	Stage string `mapstructure:"stage"`
+	// ReferenceURL can optionally give the url of the feature_gate
+	ReferenceURL string `mapstructure:"reference_url"`
+	// FromVersion optional field which gives the release version from which the gate has been given the current stage
+	FromVersion string `mapstructure:"from_version"`
+	// ToVersion optional field which gives the release version till which the gate the gate had the given lifecycle stage
+	ToVersion string `mapstructure:"to_version"`
+	// FeatureGateName name of the feature gate
+	FeatureGateName featureGateName `mapstructure:"-"`
+}
+
+func (f *featureGate) validate(parser *confmap.Conf) error {
+	if !parser.IsSet("id") {
+		return errors.New("missing required field: `id`")
+	}
+	var err []error
+	if !idRegexp.MatchString(fmt.Sprintf("%v", parser.Get("id"))) {
+		err = append(err, fmt.Errorf("invalid character(s) in ID"))
+	}
+	if !versionRegexp.MatchString(fmt.Sprintf("%v", parser.Get("from_version"))) {
+		err = append(err, fmt.Errorf("invalid character(s) in from_version"))
+	}
+	if !versionRegexp.MatchString(fmt.Sprintf("%v", parser.Get("to_version"))) {
+		err = append(err, fmt.Errorf("invalid character(s) in to_version"))
+	}
+	if !referenceURLRegexp.MatchString(fmt.Sprintf("%v", parser.Get("reference_url"))) {
+		err = append(err, fmt.Errorf("invalid character(s) in reference_url"))
+	}
```

Comment:
All of these should check first where the field is set, otherwise they are effectively required, aren't they?

For example, instead of:

```go
if !versionRegexp.MatchString(fmt.Sprintf("%v", parser.Get("from_version"))) {
	err = append(err, fmt.Errorf("invalid character(s) in from_version"))
}
```

we should do something similar to this

```go
if parser.IsSet("from_version") && !versionRegexp.MatchString(fmt.Sprintf("%v", parser.Get("from_version"))) {
	err = append(err, fmt.Errorf("invalid character(s) in from_version"))
}
```

Reply:
done